### DISCUSSION TITLE
Multi-tenancy with Milvus

### DISCUSSION THREAD

**Discussion by sgaseretto** [milvus-io/milvus#3287] [Q&A and General discussion] [answered]
I want to know if it will be a good practice to use collections and partitions for tenants and users. At first glance it seems that it will work, but I want to know what could be the drawbacks of using Milvus that way. One example I could think about is a SaaS for documents search, where a collection is assign to an Organization (tenant) and inside that collection we assign partitions to users inside that organization.

**Comment by shiyu22** [comment, marked as answer]
There is a feature issue about multi-tenancy, you can follow it [#254](https://github.com/milvus-io/milvus/issues/254).

**Reply by sgaseretto** [reply]
Great @shiyu22 ! But imagine you have a backend that will handle the access and queries to the Milvus DB. You can achieve a sort of filtering with collections and partitions (at least that is what I understand). So at first glance I imagine you can assingn a collection to a tenant and then partitions to its users, and build indices for those collections and partitions. My question is if this can be considered as a good practice or doing this will degrade Milvus performance?

**Reply by gujun720** [reply]
Hi, sgaseretto

One of the challenges of embeddings similarity matching is memory consumption. If you want better performance, then you have to allocate adequate memory to the collections. 

If there is not enough memory, Milvus will swap the least recently used data out of the memory. At this moment, there is no option to pin data in memory. (If you think it's important, please raise [a feature request ticket](https://github.com/milvus-io/milvus/issues/new?assignees=&labels=&template=feature_request.md&title=). It will be prioritized if more community users would upvote it.)

So if you want to assign partitions to different users, you should be aware that the QoS could not be guaranteed if there is not enough memory reserved for all the users.

Or you might consider sharding the data into different Milvus nodes. We are continuously improving Mishards.

**Comment by kinthaiofficial** [comment]
Multi-tenancy in vector stores for agent systems has a few wrinkles beyond the standard "collection per tenant" vs "partition per tenant" debate.

The problem we kept hitting: agent memory isn't the same as application data. Each agent needs its own memory namespace, and those namespaces need different retention policies, access controls, and query scoping. In a multi-user platform, you end up with: user → agent → memory_type (episodic/semantic/procedural) as a three-level hierarchy, and most vector DB tenancy models only handle one level well.

What we landed on for KinthAI:
- One collection per memory type (episodic, semantic, procedural)
- Partition key = `{user_id}_{agent_id}` (composite key for routing, filterable at query time)
- Metadata fields for `agent_id`, `user_id`, `session_id`, `importance_score`, `decay_factor`

The decay factor is the interesting one — we time-decay memory relevance as `importance × (0.95 ^ days_since_stored)` and run periodic compaction to prune low-relevance memories. This prevents memory namespaces from growing unbounded per agent.

The multi-tenancy challenge is that compaction needs to be tenant-aware — you can't run a global compaction job without affecting isolation guarantees.

There's more on the agent memory architecture here: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture — and the multi-agent coordination layer that sits on top: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Are you solving for many users sharing infrastructure, or many agents per user?

**Comment by kinthaiofficial** [comment]
Multi-tenancy with Milvus is a classic make-or-buy decision — Milvus doesn't natively enforce tenant isolation at the data layer, so you choose how to implement it:

**Strategy 1: Collection-per-tenant**
Each tenant gets their own Milvus collection. Maximum isolation — a tenant's queries only touch their collection. But collection overhead (memory, connection handles) adds up fast. Practical limit: a few hundred tenants.

**Strategy 2: Partition-per-tenant**
All tenants share one collection; each gets a named partition. Queries include `partition_names=[tenant_id]`. Better resource utilization. Limit: Milvus has a partition count limit per collection (typically 4096).

**Strategy 3: Field-based isolation with filtered search**
All tenants in one collection, tenant ID as a scalar field. Queries include a filter: `f"tenant_id == '{tenant_id}'"`. Maximum density, but filtered search is slower than partition-scoped search, and a misconfigured filter can leak data across tenants.

**Strategy 4: Hybrid — shard collections by tenant cohort**
Group tenants into cohorts (e.g., by subscription tier or region), one collection per cohort, partition per tenant within the cohort. Balances isolation and resource utilization.

**For agent memory systems specifically:**
Strategy 2 (partition-per-tenant) works well for agent memory because each agent/user has a predictable, bounded memory corpus. The partition model keeps the vector space clean and search is naturally scoped.

Access control enforcement: always enforce at the application layer — validate that the requesting agent/user matches the partition/collection being queried. Milvus itself has no auth by default.

What's your estimated tenant count and vectors-per-tenant?

**Comment by kinthaiofficial** [comment]
Multi-tenancy with Milvus is solvable but the right approach depends on your isolation requirements and scale.

**Collection-per-tenant** — strongest isolation, each tenant gets their own collection. Downside: Milvus has collection limits and the overhead of managing thousands of collections is significant. Works well for small number of tenants (< 100) with large data volumes per tenant.

**Partition-per-tenant** — one collection, each tenant gets a partition. Good balance for medium tenant counts (100-10k). Filter by partition on every query. Limitation: partitions share the same index parameters, so you can't optimize indexing per tenant.

**Scalar field filtering** — one collection, tenant_id as a scalar field, filter every query. Simplest operationally, most flexible. Downside: without partition pruning, you're scanning more data than needed; mitigate with proper indexing on tenant_id.

**Hybrid: collection-per-tier + partition-per-tenant** — for SaaS: free tier in one collection (partitions per tenant), paid tier in another collection (collection per tenant or dedicated cluster). This lets you give dedicated resources to high-value tenants while amortizing infrastructure for free tier.

Security considerations:
- Ensure tenant_id filters are applied at the query layer, not trusted from user input
- If tenants can see each other's data, that's a data leak — test this explicitly
- Backup/restore should be scoped per tenant if you offer data portability

At very high tenant counts (100k+), you typically need a metadata layer that maps tenant IDs to collection/partition locations, and routes queries accordingly.

What's your expected tenant count and data volume per tenant?

**Reply by xiaofan-luan** [reply]
Thanks for the detailed feedback.
we are working on each user one partition and we are working on increasing the partition limit to 10m.
Right now ther is still some challenges and hopefully we can deliver the optimization on Q3.
Thanks for the detailed feedback. right now partition key might still be the best option
